File: src/main/java/com/freundtech/minecraft/oneslotserver/Playerinfo.java
package com.freundtech.minecraft.oneslotserver;

import java.io.IOException;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.Date;

import org.bukkit.configuration.file.FileConfiguration;
import org.bukkit.configuration.file.YamlConfiguration;
import org.bukkit.entity.Player;

public class Playerinfo {	
	public Player player;
	
	FileConfiguration userconfig;
	Path configpath;
	
	public int timeleft;
	public long joinedAt;
	public long firstJoin;
	
	
	public Playerinfo(Player player, OneSlotServer plugin) {		
		configpath = Paths.get(plugin.getDataFolder().getAbsolutePath(), "players", player.getUniqueId().toString() + ".yml");
		userconfig = YamlConfiguration.loadConfiguration(configpath.toFile());
		
		this.player = player;
		this.joinedAt = new Date().getTime()/1000;
		this.firstJoin = userconfig.getLong("first_join", this.joinedAt);
		this.timeleft = userconfig.getInt("time_left", plugin.playtime);
	}
	
	public void save() {
		userconfig.set("first_join", this.firstJoin);
		userconfig.set("time_left", this.timeleft - (new Date().getTime()/1000 - this.joinedAt));
		try {
			userconfig.save(configpath.toFile());
		} catch (IOException e) {
			e.printStackTrace();
		}
	}
}

File: src/main/kotlin/com/freundtech/minecraft/oneslotserver/PlayerInfo.kt
package com.freundtech.minecraft.oneslotserver

import com.freundtech.minecraft.oneslotserver.util.delegate
import com.freundtech.minecraft.oneslotserver.util.currentTime
import org.bukkit.configuration.file.YamlConfiguration
import org.bukkit.entity.Player
import java.nio.file.Paths
import java.util.*
import kotlin.collections.HashMap

class PlayerInfo(uuid: UUID) {
    private val plugin = OneSlotServer.instance

    private val configPath = Paths.get(plugin.dataFolder.absolutePath, "players", "$uuid.yml")
    private val userConfig = YamlConfiguration.loadConfiguration(configPath.toFile())

    val joinedAt = currentTime()
    var timeLeft = userConfig.getInt("time_left", plugin.playTime)
    var firstJoin by userConfig.delegate("first_join", joinedAt)

    fun save() {
        userConfig.set("time_left", this.timeLeft - (currentTime() - this.joinedAt))
        userConfig.save(configPath.toFile())
    }
}

val playerCache = HashMap<UUID, PlayerInfo>()

val Player.playerInfo: PlayerInfo
    get() = playerCache.getOrPut(uniqueId) { PlayerInfo(uniqueId) }
